### PR TITLE
Simplify the GUS reset register logic and fix DMA IRQ handling

### LINKED ISSUE #3645: King's Quest IV: The Perils of Rosella Ultimate trips some debug asserts
King's Quest IV: The Perils of Rosella Ultimate (eXoDOS has it) crashes in with the debug build.

It's just a "safety check" style of assert, the release build runs fine.

The root cause seems to be the GUS and the SB emulation both being enabled at the same time, and the game just doing something weird...

Not sure if this is an actual problem we need to worry about at all, but documenting it anyway. Giving it the "bug" label as there's no better option.

image

### PR DESCRIPTION
**Pull request by interloper98**
# Description

This is just a general bug fix that closes #3645, #3649, #3963.

Changes are split up as small as possible to help catch any future regressions.

# Release notes

Fixed a regression that prevented the GUS from playing audio in _HELL: A Cyberpunk Thriller_ and in the 1997 demo _'Atlantis, Deep Like A Sea'_.

# Manual testing

I've tested plenty of demos, mod player, GUS-MIDI games, and GUS tracker games, and haven't hit any regressions or issues. I searched the issue tracker for closed GUS tickets and tested those too (Star Control II, Star Trek Next Generation, Solty's, Warcraft 2, and they're working, too).

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [ ] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [x] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [x] provided the release notes draft (for significant user-facing changes).
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: src/hardware/gus.cpp [hunk 1]
```diff
@@ -197,6 +195,88 @@ class Voice {
...
 static void GUS_DMA_Event(uint32_t val);
 
+// DRAM DMA Control Register (41h), section 2.6.1.1, page 12, of the UltraSound
+// Software Development Kit (SDK) Version 2.22
+//
+// This register is wrapped in a class because its state differs when written
+// versus read. When written, bit 6 indicates if the samples are 16-bit or
+// 8-bit. However when read, bit 6 indicates if a terminal count IRQ is pending.
+// There's no way for the application to read back the 16-bit sample indicator;
+// it's a one-way write into the GUS.
```
Comment (johnnovak) [inline src/hardware/gus.cpp:205]: 🥇 Very nice explanatory comment!

File: src/hardware/gus.cpp [hunk 2]
```diff
@@ -197,6 +195,88 @@ class Voice {
...
+struct DmaControlRegister {
+public:
+	constexpr void Reset() noexcept
+	{
+		reg.data          = {};
+		are_samples_16bit = {};
+	}
+	constexpr void Set(const uint8_t data) noexcept
+	{
+		// Save our existing pending TC IRQ bit
```
Comment (johnnovak) [inline src/hardware/gus.cpp:216]: Maybe spell out "TC" in the comment.

File: src/hardware/gus.cpp [hunk 3]
```diff
@@ -997,11 +1078,9 @@ bool Gus::PerformDmaTransfer()
 	}
 
 	if (dma_channel->has_reached_terminal_count) {
-		// We've hit the terminal count, so enable that bit
-		dma_ctrl |= DMA_TC_STATUS_BITMASK;
+		dma_control_register.SetPendingTerminalCountIrq();
 
-		// Raise the TC irq if needed
-		if ((dma_ctrl & 0x20) != 0) {
+		if (dma_control_register.WantsIrqOnTerminalCount()) {
```
Comment (johnnovak) [inline src/hardware/gus.cpp:1083]: This is great with the descriptive method names—top notch stuff 👍🏻

**Review by johnnovak — APPROVED** [review]
Solid work @interloper98, as always 🥳 🚀 

It seems you've put lots of effort into testing this, and as this fixes some GUS regressions, I'm thinking it could be part of 0.82.0 if your confidence levels are high? What do you think?

**Comment by johnnovak** [comment]
@interloper98 You mentioned https://github.com/dosbox-staging/dosbox-staging/issues/3649 too. Does this then fix not only the HELL: A Cyberpunk Thriller regression but also some other long-standing bugs? If so, should we call them out in the release notes?

File: src/hardware/gus.cpp (hunk 2 above)
Comment (interloper98) [reply src/hardware/gus.cpp]: Good news - in the latest update, I was able to drop the class wrapper and use the bit-view register directly; so this `Set()` went away along with the need to save/load temp variables.

File: src/hardware/gus.cpp (hunk 3 above)
Comment (interloper98) [reply src/hardware/gus.cpp]: Thanks! It's wordy :laughing: but self documenting (and matches the docs description), so :man_shrugging: :smile:

**Comment by interloper98** [comment]
> Solid work @interloper98, as always 🥳 🚀
> 
> It seems you've put lots of effort into testing this

Thanks @johnnovak - I started looking into that 'Hell: A Cyberpunk Thriller' regression (which generates a crash in debug builds). I thought it would be something minor.. but was more involved :laughing: 

'Hell' doesn't use IRQs to signal end-of-transfer, and instead polls the GUS's 'terminal count pending' bit. Comparing the docs to the code showed Staging's GUS DMA handling was always generating IRQs, even if the application didn't want them.

>  this fixes some GUS regressions, I'm thinking it could be part of 0.82.0 if your confidence levels are high? What do you think?

I think so too. I've done a lot of testing with lots of different games and programs, and recently did some longer stress testing runs, and it's running good. Even though this PR just fixes a couple regressions, the DMA IRQ handling is now compliant to the docs, without using kludges or magic logic (all of which helps my confidence).

> You mentioned https://github.com/dosbox-staging/dosbox-staging/issues/3649 too. Does this then fix not only the HELL: A Cyberpunk Thriller regression but also some other long-standing bugs? If so, should we call them out in the release notes?

Yes - it fixes that King's Quest IV debug crash you caught and that 1997 demo 'Atlantis, Deep Like A Sea' seems to play fine too (I'll mention that in the release notes):

https://github.com/user-attachments/assets/a96b6ad8-073b-4132-9890-77121c449ce8

I double checked the Warcraft II issue mention mentioned in #3649 (_"configure the game to use the Gravis Ultrasound for both music and sound, the GUS support code for digital sound will cause the game to hang if music is playing at the same time within the main menu"_) and it doesn't crash in the main menu. It's possible this has worked for some time though.

**Comment by johnnovak** [comment]
Great stuff @interloper98, thanks for confirming the above and for all your testing. 🏅 

Yeah, let's make this part of 0.82.0 then, it doesn't seem risky.

**Comment by johnnovak** [comment]
@Python-Exoproject Might be interesting to you — see the release notes section in the top description.

**Pull request merged**